Throwing java.nio.BufferUnderflowException during commands deserialization

During experimenting with the sample project (`sbt 'project example' 'runMain aecor.example.App'`), it turned out that when you send a command to aggregate (in particular, to `TransactionAggregate`), you get such error:
```
aecor.encoding.WireProtocol$Decoder$DecodingFailure: null
	at aecor.encoding.WireProtocol$Decoder$DecodingResult$.fromTry(WireProtocol.scala:42)
	at aecor.example.domain.transaction.TransactionAggregate$$anon$1$$anon$10.decode(TransactionAggregate.scala:38)
	at aecor.runtime.akkapersistence.AkkaPersistenceRuntimeActor.aecor$runtime$akkapersistence$AkkaPersistenceRuntimeActor$$handleCommand(AkkaPersistenceRuntimeActor.scala:174)
	at aecor.runtime.akkapersistence.AkkaPersistenceRuntimeActor$$anonfun$receiveCommand$1.applyOrElse(AkkaPersistenceRuntimeActor.scala:160)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:517)
	at aecor.runtime.akkapersistence.AkkaPersistenceRuntimeActor.akka$persistence$Eventsourced$$super$aroundReceive(AkkaPersistenceRuntimeActor.scala:75)
	at akka.persistence.Eventsourced$$anon$1.stateReceive(Eventsourced.scala:681)
	at akka.persistence.Eventsourced$class.aroundReceive(Eventsourced.scala:192)
	at aecor.runtime.akkapersistence.AkkaPersistenceRuntimeActor.aroundReceive(AkkaPersistenceRuntimeActor.scala:75)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:527)
	at akka.actor.ActorCell.invoke(ActorCell.scala:496)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.nio.BufferUnderflowException: null
	at java.nio.HeapByteBuffer.get(HeapByteBuffer.java:151)
	at java.nio.ByteBuffer.get(ByteBuffer.java:715)
	at boopickle.DecoderSize.readByteArray(CodecSize.scala:235)
	at boopickle.BasicPicklers$$anon$5.unpickle(Pickler.scala:379)
	at boopickle.BasicPicklers$BigDecimalPickler$.unpickle(Pickler.scala:129)
	at boopickle.BasicPicklers$BigDecimalPickler$.unpickle(Pickler.scala:120)
	at boopickle.Pickler$$anon$1.unpickle(Pickler.scala:19)
	at boopickle.PicklerHelper$class.read(Pickler.scala:49)
	at boopickle.Default$.read(Default.scala:144)
	at boopickle.TuplePicklers$$anon$4.unpickle(TuplePicklers.scala:17)
	at boopickle.TuplePicklers$$anon$4.unpickle(TuplePicklers.scala:15)
	at boopickle.UnpickleState.unpickle(Pickler.scala:586)
	at aecor.example.domain.transaction.TransactionAggregate$$anon$1$$anon$10$$anonfun$decode$1.apply(TransactionAggregate.scala:39)
	at aecor.example.domain.transaction.TransactionAggregate$$anon$1$$anon$10$$anonfun$decode$1.apply(TransactionAggregate.scala:38)
	at scala.util.Try$.apply(Try.scala:192)
```
I didn't dive into this deeply, but I noticed that such error occurs if you perform deserialization in several steps using `UnpickleState`. If you do it at once like this:
```scala
val (command, fromAccountId, toAccountId, amount) = boopickle.UnpickleImpl[Tuple4[String, From[AccountId], To[AccountId], Amount]].fromBytes(bytes)
```
Then, it works ok.

--- a/modules/boopickle-wire-protocol/src/main/scala/aecor/macros/boopickleWireProtocol.scala
+++ b/modules/boopickle-wire-protocol/src/main/scala/aecor/macros/boopickleWireProtocol.scala
@@ -122,8 +122,10 @@ object BoopickleWireProtocolMacro {
 
             final val decoder: _root_.aecor.encoding.WireProtocol.Decoder[aecor.data.PairE[$unifiedInvocation, aecor.encoding.WireProtocol.Encoder]] =
               new _root_.aecor.encoding.WireProtocol.Decoder[aecor.data.PairE[$unifiedInvocation, aecor.encoding.WireProtocol.Encoder]] {
+
                 override def decode(bytes: _root_.java.nio.ByteBuffer): _root_.aecor.encoding.WireProtocol.Decoder.DecodingResult[aecor.data.PairE[$unifiedInvocation, aecor.encoding.WireProtocol.Encoder]] = {
-                  val state = _root_.boopickle.UnpickleState(bytes)
+                  _root_.aecor.encoding.WireProtocol.Decoder.DecodingResult.fromTry(scala.util.Try {
+                  val state = _root_.boopickle.UnpickleState(bytes.order(_root_.java.nio.ByteOrder.LITTLE_ENDIAN))
                   state.unpickle[String] match {
                      ..case ${
         val cases = traitStats.map {
@@ -131,40 +133,34 @@ object BoopickleWireProtocolMacro {
             val arglist = (1 to params.size).map(i => s"_$i").map(x => q"args.${Term.Name(x)}")
             val tupleTpeBase = Type.Name(s"Tuple${params.size}")
             val nameLit = Lit.String(name.value)
-            val stats = q"""
-                            aecor.encoding.WireProtocol.Decoder.DecodingResult.fromTry(scala.util.Try {
-                              val args = state.unpickle[$tupleTpeBase[..${params.map { case param"$n:${Some(tpe)}" => toType(tpe) }}]]
-                              val invocation = new io.aecor.liberator.Invocation[$unifiedBase, $out] {
-                                final override def invoke[F[_]](mf: $unifiedBase[F]): F[$out] =
-                                  mf.$name(..$arglist)
-                                final override def toString = {
-                                  val name = $nameLit
-                                  s"$$name$$args"
-                                }
-                              }
-                              aecor.data.PairE(invocation, aecor.encoding.WireProtocol.Encoder.instance[$out](Pickle.intoBytes))
-                            })
-                        """
-            p"case $nameLit => $stats"
+            p"""case $nameLit =>
+                  val args = state.unpickle[$tupleTpeBase[..${params.map { case param"$n:${Some(tpe)}" => toType(tpe) }}]]
+                  val invocation = new io.aecor.liberator.Invocation[$unifiedBase, $out] {
+                    final override def invoke[F[_]](mf: $unifiedBase[F]): F[$out] =
+                      mf.$name(..$arglist)
+                    final override def toString: String = {
+                      val name = $nameLit
+                      s"$$name$$args"
+                    }
+                  }
+                  aecor.data.PairE(invocation, aecor.encoding.WireProtocol.Encoder.instance[$out](Pickle.intoBytes))
+             """
           case q"def $name: ${someF: Type.Name}[$out]" if someF.value == theF.value =>
-            val stats = q"""
-                            aecor.encoding.WireProtocol.Decoder.DecodingResult.fromTry(scala.util.Try {
-                              val invocation = new io.aecor.liberator.Invocation[$unifiedBase, $out] {
-                                final override def invoke[F[_]](mf: $unifiedBase[F]): F[$out] =
-                                  mf.$name
-                              }
-                              aecor.data.PairE(invocation, aecor.encoding.WireProtocol.Encoder.instance[$out](Pickle.intoBytes))
-                            })
-                        """
-            val nameLit = Lit.String(name.value)
-            p"case $nameLit => $stats"
+            val nameLit =  Lit.String(name.value)
+            p"""case $nameLit =>
+                  val invocation = new io.aecor.liberator.Invocation[$unifiedBase, $out] {
+                    final override def invoke[F[_]](mf: $unifiedBase[F]): F[$out] = mf.$name
+                    final override def toString: String = $nameLit
+                  }
+                  aecor.data.PairE(invocation, aecor.encoding.WireProtocol.Encoder.instance[$out](Pickle.intoBytes))
+             """
           case other =>
             abort(s"Illegal method [$other]")
         }
-        cases :+ p"""case other => scala.util.Left(aecor.encoding.WireProtocol.Decoder.DecodingFailure(s"Unknown type tag $$other"))"""
+        cases :+ p"""case other => throw aecor.encoding.WireProtocol.Decoder.DecodingFailure(s"Unknown type tag $$other")"""
       }
                   }
-
+                  })
                 }
               }
           }
@@ -174,7 +170,7 @@ object BoopickleWireProtocolMacro {
     val newCompanion = companion match {
       case Some(c) =>
         val oldTemplStats = c.templ.stats.getOrElse(Nil)
-        c.copy(templ = c.templ.copy(stats = Some(companionStats ++ oldTemplStats)))
+        c.copy(templ = c.templ.copy(stats = Some(oldTemplStats ++ companionStats)))
       case None =>
         q"object ${Term.Name(typeName.value)} { ..$companionStats }"
 

--- a/modules/example/src/main/resources/logback.xml
+++ b/modules/example/src/main/resources/logback.xml
@@ -6,7 +6,7 @@
         </encoder>
     </appender>
 
-    <root level="OFF">
+    <root level="DEBUG">
         <appender-ref ref="STDOUT" />
     </root>
     <logger name="org.apache.kafka.clients.consumer.KafkaConsumer" level="WARN" />

